pallets/foreign-investments/src/impls/mod.rs

```diff
@@ -0,0 +1,1042 @@
+};
+
+mod invest;
+mod redeem;
+
+// Handles the second stage of updating investments. Whichever (potentially
+// async) code path of the first stage concludes it (partially) should call
+// `Swap::Config::SwapNotificationHandler::notify_status_update(swap_order_id,
+// swapped_amount)`.
+impl<T: Config> StatusNotificationHook for Pallet<T> {
+	type Error = DispatchError;
+	type Id = T::TokenSwapOrderId;
+	type Status = SwapOf<T>;
+
+	fn notify_status_change(
+		id: T::TokenSwapOrderId,
+		status: SwapOf<T>,
+	) -> Result<(), DispatchError> {
+		let info = ForeignInvestmentInfo::<T>::get(id).ok_or(Error::<T>::InvestmentInfoNotFound)?;
+		let reason = info
+			.last_swap_reason
+			.ok_or(Error::<T>::TokenSwapReasonNotFound)?;
+
+		match reason {
+			TokenSwapReason::Investment => {
+				let pre_state = InvestmentState::<T>::get(&info.owner, info.id).unwrap_or_default();
+				let post_state = pre_state
+					.transition(InvestTransition::FulfillSwapOrder(status))
+					.map_err(|e| {
+						// Inner error holds finer granularity but should never occur
+						log::debug!("ForeignInvestment state transition error: {:?}", e);
+						Error::<T>::from(InvestError::FulfillSwapOrder)
+					})?;
+				Pallet::<T>::apply_invest_state_transition(&info.owner, info.id, post_state)
+			}
+			TokenSwapReason::Redemption => {
+				let pre_state = RedemptionState::<T>::get(&info.owner, info.id).unwrap_or_default();
+				let post_state = pre_state
+					.transition(RedeemTransition::FulfillSwapOrder(status))
+					.map_err(|e| {
+						// Inner error holds finer granularity but should never occur
+						log::debug!("ForeignInvestment state transition error: {:?}", e);
+						Error::<T>::from(RedeemError::FulfillSwapOrder)
+					})?;
+				Pallet::<T>::apply_redeem_state_transition(&info.owner, info.id, post_state)
+			}
+		}
+	}
+}
+
+impl<T: Config> ForeignInvestment<T::AccountId> for Pallet<T> {
+	type Amount = T::Balance;
+	type CollectInvestResult = ExecutedForeignCollectInvest<T::Balance>;
+	type CurrencyId = T::CurrencyId;
+	type Error = DispatchError;
+	type InvestmentId = T::InvestmentId;
+
+	#[transactional]
+	fn increase_foreign_investment(
+		who: &T::AccountId,
+		investment_id: T::InvestmentId,
+		amount: T::Balance,
+		return_currency: T::CurrencyId,
+		pool_currency: T::CurrencyId,
+	) -> Result<(), DispatchError> {
+		let pre_state = InvestmentState::<T>::get(who, investment_id).unwrap_or_default();
+
+		// NOTE: This adds one db read but let's be safe for the MVP, can hopefully be
+		// removed before deploying to production
+		let unprocessed_invest_amount = T::Investment::investment(who, investment_id)?;
+		ensure!(
+			unprocessed_invest_amount == pre_state.get_investing_amount(),
+			DispatchError::Corruption
+		);
+
+		let post_state = pre_state
+			.transition(InvestTransition::IncreaseInvestOrder(Swap {
+				currency_in: pool_currency,
+				currency_out: return_currency,
+				amount,
```

Comment:
I still getting a mess 😆 with those names. From the `Swap` docs I understand I give `currency_out` and I get after the swap the desired `currency_in`. So here I'm giving to the swap the return currency (although by the name it seems like it's what I want to get) and I'm getting `pool_currency`. Is my explanation ok?

Reply:
Thanks for this great explanation!! 🙌🏻 

I'll think if I find a better name during the review process once I get more knowledge, but for now, I think `foreign` makes sense.